### PR TITLE
docs(pubsub): add a subscriber quickstart

### PR DESCRIPTION
**Pull request by alevenberg**
Create a subscriber executable for the Pub/Sub quickstart

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12053)

### CODE REVIEW

File: google/cloud/pubsub/README.md [hunk 1]
```diff
@@ -32,30 +32,19 @@ int main(int argc, char* argv[]) try {
...
+  std::cout << "Hello World published with id=" << *id << std::endl;
 
   return 0;
 } catch (google::cloud::Status const& status) {
-  std::cerr << "google::cloud::Status thrown: " << status << "\n";
+  std::cerr << "google::cloud::Status thrown: " << status << std::endl;
   return 1;
 }
 ```
-
```
Comment (coryan) [inline google/cloud/pubsub/README.md:50]: You probably lost the rest of this file by accident, please restore.

File: google/cloud/pubsub/quickstart/README.md [hunk 2]
```diff
@@ -51,7 +51,7 @@ authentication preferences, though if the environment variable is not set, the
 library searches for a credentials file in the same location as the [Cloud
 SDK](https://cloud.google.com/sdk/). For more information about *Application
 Default Credentials*, see
-https://cloud.google.com/docs/authentication/production
+<https://cloud.google.com/docs/authentication/production>
```
Comment (coryan) [inline google/cloud/pubsub/quickstart/README.md:54]: Any reason to use angle brackets here?

File: google/cloud/pubsub/quickstart/quickstart.cc [hunk 3]
```diff
@@ -27,18 +27,19 @@ int main(int argc, char* argv[]) try {
...
-  std::cout << "Hello World published with id=" << *id << "\n";
+  std::cout << "Hello World published with id=" << *id << std::endl;
 
   return 0;
 } catch (google::cloud::Status const& status) {
-  std::cerr << "google::cloud::Status thrown: " << status << "\n";
+  std::cerr << "google::cloud::Status thrown: " << status << std::endl;
   return 1;
 }
 //! [END pubsub_quickstart_publisher] [all]
```
Comment (coryan) [inline google/cloud/pubsub/quickstart/quickstart.cc:45]: cloud.google.com probably has code referencing the region tag in this file.  See my other comments regarding moving region tags.

File: google/cloud/pubsub/quickstart/quickstart.cc [hunk 4]
```diff
@@ -27,18 +27,19 @@ int main(int argc, char* argv[]) try {
...
+
   auto publisher = pubsub::Publisher(
       pubsub::MakePublisherConnection(pubsub::Topic(project_id, topic_id)));
   auto id =
       publisher
           .Publish(pubsub::MessageBuilder{}.SetData("Hello World!").Build())
           .get();
   if (!id) throw std::move(id).status();
-  std::cout << "Hello World published with id=" << *id << "\n";
+  std::cout << "Hello World published with id=" << *id << std::endl;
```
Comment (coryan) [inline google/cloud/pubsub/quickstart/quickstart.cc:38]: `std::endl` both prints a newline and flushes the output.  You almost certainly do not need to flush the output in these examples?

File: google/cloud/pubsub/quickstart/subscriber_quickstart.cc [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            std::cout << "Received message " << m << std::endl;
+            std::move(h).ack();
+          })
+          .then([](auto f) { throw f.get(); });
+
+  std::cout << "Waiting for messages on " + subscription_id + "..."
+            << std::endl;
+
+  // Blocks indefinitely, unless an exception is thrown, since Subscribe should
+  // be used in a long running operation.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/pubsub/quickstart/subscriber_quickstart.cc:47]: Also, it is possible to stop early by calling `session.cancel()`.

File: google/cloud/pubsub/quickstart/subscriber_quickstart.cc [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  auto subscriber = pubsub::Subscriber(pubsub::MakeSubscriberConnection(
+      pubsub::Subscription(project_id, subscription_id)));
+
+  auto session =
+      subscriber
+          .Subscribe([&](pubsub::Message const& m, pubsub::AckHandler h) {
+            std::cout << "Received message " << m << std::endl;
+            std::move(h).ack();
+          })
+          .then([](auto f) { throw f.get(); });
```
Comment (coryan) [inline google/cloud/pubsub/quickstart/subscriber_quickstart.cc:41]: Are you certain this is what you want to do?

**Review by coryan — CHANGES_REQUESTED** [review]

File: google/cloud/pubsub/quickstart/README.md [hunk 7]
```diff
@@ -152,4 +164,4 @@ set GRPC_DEFAULT_SSL_ROOTS_FILE_PATH=%cd%\roots.pem
 [grpc-roots-pem-bug]: https://github.com/grpc/grpc/issues/16571
 [homebrew-cmake-link]: https://formulae.brew.sh/formula/cmake
 [howto-setup-dev-workstation]: /doc/contributor/howto-guide-setup-development-workstation.md
-[pubsub-quickstart-link]: https://cloud.google.com/pubsub/docs/quickstart-console
+[pubsub-quickstart-link]: https://cloud.google.com/pubsub/docs/publish-receive-messages-client-library
```
Comment (devbww) [inline google/cloud/pubsub/quickstart/README.md:167]: Note that moving the snippet tag from one file to another will require some matching change to the docs.

File: google/cloud/pubsub/quickstart/quickstart.cc (hunk 4 above)
Comment (alevenberg) [reply google/cloud/pubsub/quickstart/quickstart.cc]: Removed the endls across the quickstart

File: google/cloud/pubsub/README.md (hunk 1 above)
Comment (alevenberg) [reply google/cloud/pubsub/README.md]: Restored

File: google/cloud/pubsub/quickstart/README.md (hunk 7 above)
Comment (alevenberg) [reply google/cloud/pubsub/quickstart/README.md:166]: What do you mean matching change to the docs?

File: google/cloud/pubsub/quickstart/subscriber_quickstart.cc (hunk 6 above)
Comment (alevenberg) [reply google/cloud/pubsub/quickstart/subscriber_quickstart.cc]: No...

I want it to throw an exception if the Pub/Sub service returns one.

Added a check to see when the session completes if the status is ok or not.

File: google/cloud/pubsub/quickstart/subscriber_quickstart.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/quickstart/subscriber_quickstart.cc]: Changed the wording.

File: google/cloud/pubsub/quickstart/CMakeLists.txt [hunk 8]
```diff
@@ -27,6 +27,8 @@ else ()
     set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>DLL")
 endif ()
 
-# Once the pubsub_client package is found, define new targets.
+# Once the pubsub_client package is found, define new targets and link them.
```
Comment (coryan) [inline google/cloud/pubsub/quickstart/CMakeLists.txt:30]: maybe "... and the dependencies they link".

**Review by coryan — APPROVED** [review]
You probably need to rebase if you want the new build to pass.

File: google/cloud/pubsub/quickstart/subscriber_quickstart.cc [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    std::cout << "timeout reached, ending session\n";
+    session.cancel();
+  }
+
+  return 0;
+} catch (google::cloud::Status const& status) {
+  std::cerr << "google::cloud::Status thrown: " << status << "\n";
+  return 1;
+}
+
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/quickstart/subscriber_quickstart.cc:57]: Checkers did not like that...

[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/quickstart/subscriber_quickstart.cc]: Ah, I see now! Done.

**Pull request merged**